File: src/services/cmds.ts
import { invoke } from '@tauri-apps/api/core';

/**
 * @description: Get app settings data
 * @return {Promise<Nvmd.Setting>} Promise-Nvmd.Setting
 */
export async function getSettings() {
  return invoke<Nvmd.Setting>('read_settings');
}

/**
 * @description: Update app settings data
 * @param {Nvmd.Setting} settings app settings data
 * @return {Promise<void>} Promise-void
 */
export async function updateSettings(settings: Nvmd.Setting) {
  return invoke<void>('update_settings', { settings });
}

/**
 * @description: Get the current global node version
 * @param {boolean} fetch whether to read the latest value from the file
 * @return {Promise<string>} node version
 */
export function vCurrent(fetch: boolean = false) {
  return invoke<string>('current', { fetch });
}

/**
 * @description: Set the globally effective node version
 * @param {string} version node version number
 * @return {Promise<void>}	Promise-void
 */
export function vSetCurrent(version: string) {
  return invoke<void>('set_current', { version });
}

/**
 * @description: Get a list of all officially released versions of node
 * @param {boolean} fetch Whether to pull new data from remote services
 * @return {Promise<Nvmd.Versions>}	List of all officially released versions of node
 */
export function versionList(fetch: boolean = false) {
  return invoke<Nvmd.Versions>('version_list', { fetch });
}

/**
 * @description: Get a list of installed nodes
 * @param {boolean} fetch Whether to read the latest data
 * @return {Promise<Array<string>>} An array containing the node version number
 */
export function installedList(fetch: boolean = false) {
  return invoke<Array<string>>('installed_list', { fetch });
}

/**
 * @description: Download Node
 * @param {string} version node version
 * @param {string} arch	node architecture
 * @return {Promise<string>}	The file path where the downloaded node is saved
 */
export function installNode(version: string, arch?: string) {
  return invoke<string>('install_node', { version, arch });
}

/**
 * @description: Cancel the task of downloading node
 * @return {Promise<void>} Promise-void
 */
export function installNodeCancel() {
  return invoke<void>('install_node_cancel');
}

/**
 * @description	uninstall node
 * @param {string} version version number
 * @param {boolean} current whether the version to be uninstalled is the currently used version
 * @returns {Promise<void>} Promise<void>
 */
export function uninstallNode(version: string, current: boolean = false) {
  return invoke<void>('uninstall_node', { version, current });
}

/**
 * @description: Get project list from locale file
 * @param {boolean} fetch whether to read the latest value from the file
 * @return {Promise<Nvmd.Project[]>}
 */
export function projectList(fetch: boolean = false) {
  return invoke<Nvmd.Project[]>('project_list', { fetch });
}

/**
 * @description: Select projects
 * @return {Promise<Array<Nvmd.PInfo>>}
 */
export function selectProjects() {
  return invoke<Array<Nvmd.PInfo>>('select_projects');
}

/**
 * @description: Update projects data
 * @param {Array<Nvmd.Project>} list projects list
 * @return {Promise<void>}
 */
export function updateProjects(list: Nvmd.Project[], path?: string) {
  return invoke<void>('update_projects', { list, path });
}

/**
 * @description: Update project version
 * @param {string} path project floder path
 * @param {string} version node version
 * @return {Promise<200 | 404>}
 */
export function syncProjectVersion(path: string, version: string) {
  return invoke<200 | 404>('sync_project_version', { path, version });
}

/**
 * @description: Batch update project version
 * @param {string[]} paths project floder paths
 * @param {string} version node version
 * @return {Promise<void>}
 */
export function batchUpdateProjectVersion(paths: string[], version: string) {
  return invoke<void>('batch_update_project_version', { paths, version });
}

/**
 * @description: Get group list from locale file
 * @param {boolean} fetch whether to read the latest value from the file
 * @return {Promise<Nvmd.Group[]>}
 */
export function groupList(fetch: boolean = false) {
  return invoke<Nvmd.Group[]>('group_list', { fetch });
}

/**
 * @description: Update groups list & save to local file
 * @param {Array<Nvmd.Group>} list groups list
 * @return {Promise<void>}
 */
export function updateGroups(list: Nvmd.Group[]) {
  return invoke<void>('update_groups', { list });
}

/**
 * @description: Update group version
 * @param {string} name group name
 * @param {string} version version
 * @return {Promise<void>}
 */
export function updateGroupVersion(name: string, version: string) {
  return invoke<void>('update_group_version', { name, version });
}

File: src-tauri/src/cmds.rs
use anyhow::Result;
use dark_light::{detect as detect_system_theme, Mode as SystemTheme};
use std::{path::PathBuf, process::Command};
use tauri_plugin_window_state::{AppHandleExt, StateFlags};

use crate::{
    config::{Config, Group, ISettings, ISettingsResponse, NVersion, Project},
    core::{configuration, group, handle, node, project},
    utils::dirs,
    wrap_err,
};

type CmdResult<T = ()> = Result<T, String>;

/// get current version
#[tauri::command]
pub fn current(fetch: Option<bool>) -> CmdResult<Option<String>> {
    wrap_err!(node::get_current(fetch))
}

/// set current version
#[tauri::command]
pub async fn set_current(version: Option<String>) -> CmdResult<()> {
    wrap_err!(node::set_current(version).await)
}

/// fetch node version list
#[tauri::command]
pub async fn version_list(fetch: Option<bool>) -> CmdResult<Option<Vec<NVersion>>> {
    wrap_err!(node::get_version_list(fetch).await)
}

/// read node installed version list
#[tauri::command]
pub async fn installed_list(fetch: Option<bool>) -> CmdResult<Option<Vec<String>>> {
    wrap_err!(node::get_installed_list(fetch).await)
}

/// read settings
#[tauri::command]
pub async fn read_settings() -> CmdResult<ISettingsResponse> {
    let setting_data = Config::settings().latest_ref().clone();
    Ok(ISettingsResponse::from(*setting_data))
}

/// update settings
#[tauri::command]
pub async fn update_settings(payload: ISettings) -> CmdResult<()> {
    let settings = Config::settings().latest_ref().clone();
    let locale = settings.locale;
    let directory = settings.directory;

    wrap_err!({
        Config::settings()
            .draft_mut()
            .patch_settings(payload.clone())
    })?;
    Config::settings().apply();

    // refresh data when directory changes
    if directory != payload.directory {
        wrap_err!(node::get_installed_list(Some(true)).await)?;
    }
    // update system tray
    if locale != payload.locale || directory != payload.directory {
        wrap_err!(handle::Handle::update_systray_part())?;
    }

    Ok(())
}

/// install node
#[tauri::command]
pub async fn install_node(
    window: tauri::Window,
    version: Option<String>,
    arch: Option<String>,
) -> CmdResult<String> {
    wrap_err!(node::install_node(window, version, arch).await)
}

/// install node
#[tauri::command]
pub async fn install_node_cancel() -> CmdResult<()> {
    wrap_err!(node::install_node_cancel().await)
}

/// uninstall node
#[tauri::command]
pub async fn uninstall_node(version: String) -> CmdResult<()> {
    wrap_err!(node::uninstall_node(version).await)
}

/// get project list
#[tauri::command]
pub async fn project_list(fetch: Option<bool>) -> CmdResult<Option<Vec<Project>>> {
    wrap_err!(project::project_list(fetch).await)
}

/// add projects
#[tauri::command]
pub async fn select_projects(
    app_handle: tauri::AppHandle,
) -> CmdResult<Option<Vec<project::PInfo>>> {
    wrap_err!(project::select_projects(app_handle).await)
}

/// update projects
#[tauri::command]
pub async fn update_projects(list: Vec<Project>, path: Option<PathBuf>) -> CmdResult<()> {
    wrap_err!(project::update_projects(list, path).await)
}

/// update project version
#[tauri::command]
pub async fn sync_project_version(path: PathBuf, version: String) -> CmdResult<i32> {
    wrap_err!(project::sync_project_version(path, &version).await)
}

/// batch update project version
#[tauri::command]
pub async fn batch_update_project_version(paths: Vec<PathBuf>, version: String) -> CmdResult<()> {
    wrap_err!(project::batch_update_project_version(paths, version).await)
}

/// get group list
#[tauri::command]
pub async fn group_list(fetch: Option<bool>) -> CmdResult<Option<Vec<Group>>> {
    wrap_err!(group::group_list(fetch).await)
}

/// update groups
#[tauri::command]
pub async fn update_groups(list: Vec<Group>) -> CmdResult<()> {
    wrap_err!(group::update_groups(list).await)
}

/// update group version
#[tauri::command]
pub async fn update_group_version(name: String, version: String) -> CmdResult<()> {
    wrap_err!(group::update_group_version(name, version).await)
}

/// configration export
#[tauri::command]
pub async fn configration_export(
    output_path: PathBuf,
    configration: configuration::ConfigurationExport,
) -> CmdResult<()> {
    wrap_err!(configuration::configuration_export(output_path, configration).await)
}

/// configration import
#[tauri::command]
pub async fn configration_import(
    app_handle: tauri::AppHandle,
    sync: bool,
) -> CmdResult<Option<configuration::ConfigurationImport>> {
    wrap_err!(configuration::configuration_import(&app_handle, sync).await)
}

/// open project with VsCode
#[tauri::command]
pub async fn open_with_vscode(path: String) -> CmdResult<()> {
    let cmd = { Config::settings().latest_ref().coder.clone() }.unwrap();
    let mut command = Command::new(cmd);
    command.arg(&path);

    #[cfg(target_os = "windows")]
    {
        use std::os::windows::process::CommandExt;
        use windows::Win32::System::Threading::CREATE_NO_WINDOW;
        command.creation_flags(CREATE_NO_WINDOW.0);
    }

    wrap_err!(command.status())?;
    Ok(())
}

/// open app config dir
#[tauri::command]
pub fn open_config_dir() -> CmdResult<()> {
    let data_dir: PathBuf = wrap_err!(dirs::app_config_dir())?;
    wrap_err!(open::that(data_dir))
}

/// open app data dir `.nvmd`
#[tauri::command]
pub fn open_data_dir() -> CmdResult<()> {
    let data_dir: PathBuf = wrap_err!(dirs::nvmd_home_dir())?;
    wrap_err!(open::that(data_dir))
}

/// open app logs dir
#[tauri::command]
pub fn open_logs_dir() -> CmdResult<()> {
    let logs_dir: PathBuf = wrap_err!(dirs::app_logs_dir())?;
    wrap_err!(open::that(logs_dir))
}

/// open project dir with the File Explorer
#[tauri::command]
pub fn open_dir(dir: String) -> CmdResult<()> {
    wrap_err!(open::that(dir))
}

#[tauri::command]
pub fn get_system_theme() -> CmdResult<String> {
    let theme = match detect_system_theme() {
        Ok(SystemTheme::Dark) => "dark",
        Ok(SystemTheme::Light) => "light",
        Ok(SystemTheme::Unspecified) | Err(_) => "light",
    };
    Ok(theme.to_string())
}

/// restart app
#[tauri::command]
pub fn restart(app_handle: tauri::AppHandle) {
    let _ = app_handle.save_window_state(StateFlags::default());
    app_handle.restart()
}

/// exit app
#[tauri::command]
pub fn exit_app(app_handle: tauri::AppHandle) {
    app_handle.exit(0);
}
